File: 3702-longest-subsequence-with-non-zero-bitwise-xor/3702-longest-subsequence-with-non-zero-bitwise-xor.java
class Solution {
    public int longestSubsequence(int[] nums) {
        int xor = 0;
        boolean hasNonZero = false;
        
        // calculate the xor of entire array
        // if total xor != 0 then answer = n
        // otherwise 
        // if there is at least one non zero element answer = n-1
        // else answer = 0
        // we don't need to construct the subsequence
        // Key point of the problem:
        // If totalXor is 0, removing one element (non zero) makes the remaining xor equalt to that element

        for (int num : nums) {
            xor ^= num;

            if (num != 0)
                hasNonZero = true;
        }

        if (xor != 0)
            return nums.length;
        
        return hasNonZero ? nums.length - 1 : 0;
    }
}

File: 3702-longest-subsequence-with-non-zero-bitwise-xor/3702-longest-subsequence-with-non-zero-bitwise-xor.py
class Solution:
    def longestSubsequence(self, nums: list[int]) -> int:
        xor = 0
        hasNonZero = False

        for num in nums:
            xor ^= num

            if num != 0:
                hasNonZero = True

        if xor != 0:
            return len(nums)

        return len(nums) - 1 if hasNonZero else 0
